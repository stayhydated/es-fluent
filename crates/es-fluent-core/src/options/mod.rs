pub mod r#enum;
pub mod r#struct;

const DEFAULT_FL_PATH: &str = "crate::fl";
