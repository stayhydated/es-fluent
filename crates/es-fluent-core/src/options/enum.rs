use std::str::FromStr as _;

use crate::strategy::DisplayStrategy;
use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta, FromVariant};
use getset::Getters;
use strum::IntoEnumIterator as _;

/// Options for an enum field.
#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct EnumFieldOpts {
    /// The identifier of the field.
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    /// The type of the field.
    #[getset(get = "pub")]
    ty: syn::Type,
    /// Whether to skip this field.
    #[darling(default)]
    skip: Option<bool>,
    /// Whether this field is a choice.
    #[darling(default)]
    choice: Option<bool>,
}

impl EnumFieldOpts {
    /// Returns `true` if the field should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns `true` if the field is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
}

/// Options for an enum variant.
#[derive(Clone, Debug, FromVariant, Getters)]
#[darling(attributes(fluent))]
pub struct VariantOpts {
    /// The identifier of the variant.
    #[getset(get = "pub")]
    ident: syn::Ident,
    fields: darling::ast::Fields<EnumFieldOpts>,
    /// Whether to skip this variant.
    #[darling(default)]
    skip: Option<bool>,
}

impl VariantOpts {
    /// Returns `true` if the variant should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns the style of the variant's fields.
    pub fn style(&self) -> darling::ast::Style {
        self.fields.style
    }
    /// Returns the fields of the variant that are not skipped.
    pub fn fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields
            .iter()
            .filter(|field| !field.is_skipped())
            .collect()
    }
    /// Returns all fields of the variant.
    pub fn all_fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields.iter().collect()
    }
}

/// Options for an enum.
#[derive(Clone, FromDeriveInput, Getters, Debug)]
#[darling(supports(enum_unit, enum_named, enum_tuple), attributes(fluent))]
#[getset(get = "pub")]
pub struct EnumOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<VariantOpts, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumFluentAttributeArgs,
}

impl EnumOpts {
    /// Returns the variants of the enum.
    pub fn variants(&self) -> Vec<&VariantOpts> {
        match &self.data {
            darling::ast::Data::Enum(variants) => variants.iter().collect(),
            _ => unreachable!("Unexpected data type for enum"),
        }
    }
}

/// Attribute arguments for an enum.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct EnumFluentAttributeArgs {
    display: Option<String>,
    #[darling(default)]
    choice: Option<bool>,
    #[darling(default)]
    this: Option<bool>,
}

impl EnumFluentAttributeArgs {
    /// Returns the display strategy for the enum.
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::FluentDisplay
        }
    }
    /// Returns `true` if the enum is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
    /// Returns `true` if the enum should be passed as `this`.
    pub fn is_this(&self) -> bool {
        self.this.unwrap_or(false)
    }
}

/// Options for an enum that can be used as a choice.
#[derive(Clone, Debug, FromDeriveInput, Getters)]
#[darling(supports(enum_unit), attributes(fluent_choice))]
#[getset(get = "pub")]
pub struct EnumChoiceOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumChoiceAttributeArgs,
}

/// Attribute arguments for an enum that can be used as a choice.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
#[getset(get = "pub")]
pub struct EnumChoiceAttributeArgs {
    #[darling(default)]
    serialize_all: Option<String>,
}
