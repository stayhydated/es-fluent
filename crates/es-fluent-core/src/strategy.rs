use crate::options::{r#enum::EnumOpts, r#struct::StructOpts};
use strum::{EnumIter, EnumString, IntoStaticStr};

#[derive(Clone, Copy, Debug, EnumIter, EnumString, Eq, Hash, IntoStaticStr, PartialEq)]
pub enum DisplayStrategy {
    /// use es_fluent::FluentDisplay on the target enum
    #[strum(serialize = "fluent")]
    FluentDisplay,
    /// use std::fmt::Display on the target enum
    #[strum(serialize = "std")]
    StdDisplay,
}

impl From<&EnumOpts> for DisplayStrategy {
    fn from(opts: &EnumOpts) -> Self {
        opts.attr_args().display()
    }
}

impl From<&StructOpts> for DisplayStrategy {
    fn from(opts: &StructOpts) -> Self {
        opts.attr_args().display()
    }
}
