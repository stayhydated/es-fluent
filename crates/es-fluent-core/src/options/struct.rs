use std::str::FromStr as _;

use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta};
use getset::Getters;
use quote::format_ident;
use strum::IntoEnumIterator as _;

use crate::strategy::DisplayStrategy;

/// Options for a struct field.
#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct StructFieldOpts {
    /// The identifier of the field.
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    /// The type of the field.
    #[getset(get = "pub")]
    ty: syn::Type,
    /// Whether to skip this field.
    #[darling(default)]
    skip: Option<bool>,
    /// Whether this field is a default.
    #[darling(default)]
    default: Option<bool>,
    /// Whether this field is a choice.
    #[darling(default)]
    choice: Option<bool>,
}

impl StructFieldOpts {
    /// Returns `true` if the field should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }

    /// Returns `true` if the field is a default.
    pub fn is_default(&self) -> bool {
        self.default.unwrap_or(false)
    }

    /// Returns `true` if the field is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
}

/// Options for a struct.
#[derive(Clone, Debug, FromDeriveInput, Getters)]
#[darling(supports(struct_named), attributes(fluent))]
#[getset(get = "pub")]
pub struct StructOpts {
    /// The identifier of the struct.
    ident: syn::Ident,
    /// The generics of the struct.
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, StructFieldOpts>,
    #[darling(flatten)]
    attr_args: StructFluentAttributeArgs,
}

impl StructOpts {
    /// Returns the fields of the struct that are not skipped.
    pub fn fields(&self) -> Vec<&StructFieldOpts> {
        match &self.data {
            darling::ast::Data::Struct(fields) => fields
                .fields
                .iter()
                .filter(|field| !field.is_skipped())
                .collect(),
            _ => vec![],
        }
    }
}

/// Attribute arguments for a struct.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct StructFluentAttributeArgs {
    display: Option<String>,
    #[darling(default)]
    this: Option<bool>,
    /// The traits to derive on the FTL enum.
    #[getset(get = "pub")]
    #[darling(default)]
    derive: darling::util::PathList,
}
impl StructFluentAttributeArgs {
    /// Returns the display strategy for the struct.
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::FluentDisplay
        }
    }
    /// Returns `true` if the struct should be passed as `this`.
    pub fn is_this(&self) -> bool {
        self.this.unwrap_or(false)
    }
}

/// Options for a struct field.
#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent_kv))]
pub struct StructKvFieldOpts {
    /// The identifier of the field.
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    /// The type of the field.
    #[getset(get = "pub")]
    ty: syn::Type,
    /// Whether to skip this field.
    #[darling(default)]
    skip: Option<bool>,
    /// Whether this field is a default.
    #[darling(default)]
    default: Option<bool>,
}

impl StructKvFieldOpts {
    /// Returns `true` if the field should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }

    /// Returns `true` if the field is a default.
    pub fn is_default(&self) -> bool {
        self.default.unwrap_or(false)
    }
}

/// Options for a struct.
#[derive(Clone, FromDeriveInput, Getters, Debug)]
#[darling(supports(struct_named), attributes(fluent_kv))]
#[getset(get = "pub")]
pub struct StructKvOpts {
    /// The identifier of the struct.
    ident: syn::Ident,
    /// The generics of the struct.
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, StructKvFieldOpts>,
    #[darling(flatten)]
    attr_args: StructKvFluentAttributeArgs,
}

impl StructKvOpts {
    const FTL_ENUM_IDENT: &str = "Ftl";

    /// Returns the identifier of the FTL enum.
    pub fn ftl_enum_ident(&self) -> syn::Ident {
        format_ident!("{}{}", &self.ident, Self::FTL_ENUM_IDENT)
    }

    /// Returns the identifiers of the keyed FTL enums.
    pub fn keyyed_idents(&self) -> Vec<syn::Ident> {
        self.attr_args
            .clone()
            .keys
            .map_or_else(Vec::new, |keys| {
                keys.into_iter()
                    .map(|key| format_ident!("{}", key.value()))
                    .collect()
            })
            .into_iter()
            .map(|key| format_ident!("{}{}{}", &self.ident, key, Self::FTL_ENUM_IDENT))
            .collect()
    }

    /// Returns the fields of the struct that are not skipped.
    pub fn fields(&self) -> Vec<&StructKvFieldOpts> {
        match &self.data {
            darling::ast::Data::Struct(fields) => fields
                .fields
                .iter()
                .filter(|field| !field.is_skipped())
                .collect(),
            _ => vec![],
        }
    }
}

/// Attribute arguments for a struct.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct StructKvFluentAttributeArgs {
    display: Option<String>,
    #[darling(default)]
    keys: Option<Vec<syn::LitStr>>,
    #[darling(default)]
    this: Option<bool>,
    /// The traits to derive on the FTL enum.
    #[getset(get = "pub")]
    #[darling(default)]
    derive: darling::util::PathList,
}
impl StructKvFluentAttributeArgs {
    /// Returns the display strategy for the struct.
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::StdDisplay
        }
    }
    /// Returns `true` if the struct should be passed as `this`.
    pub fn is_this(&self) -> bool {
        self.this.unwrap_or(false)
    }
}
