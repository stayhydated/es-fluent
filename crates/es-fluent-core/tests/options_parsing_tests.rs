use darling::FromDeriveInput;
use es_fluent_core::options::{
    r#enum::{EnumChoiceOpts, EnumOpts},
    r#struct::{StructKvOpts, StructOpts},
};
use es_fluent_core::strategy::DisplayStrategy;
use syn::{DeriveInput, parse_quote};

#[test]
fn enum_display_strategy_default_is_fluent() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent] // no explicit display -> default for enums is FluentDisplay
        enum MyEnum {
            Unit,
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    assert_eq!(DisplayStrategy::from(&opts), DisplayStrategy::FluentDisplay);
    assert!(!opts.attr_args().is_this());
    assert!(!opts.attr_args().is_choice());
}

#[test]
fn enum_display_strategy_override_std_and_this() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(display = "std", this)]
        enum MyEnum {
            Unit,
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    assert_eq!(DisplayStrategy::from(&opts), DisplayStrategy::StdDisplay);
    assert!(opts.attr_args().is_this());
}

#[test]
fn enum_variants_and_fields_skipping_and_choice() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent]
        enum MyEnum {
            // Struct variant with one skipped field and one choice field
            Data {
                #[fluent(choice)]
                a: i32,
                #[fluent(skip)]
                b: String,
            },
            // Tuple variant with one skipped positional field
            Tuple(#[fluent(skip)] i32, String),
            // Unit variant
            Unit,
            // Entirely skipped variant (should still be parsed, filtering is done at analysis stage)
            #[fluent(skip)]
            Skipped,
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    let variants = opts.variants();

    // Find Data variant
    let data = variants
        .iter()
        .find(|v| v.ident().to_string() == "Data")
        .expect("Data variant present");
    assert!(matches!(data.style(), darling::ast::Style::Struct));
    // fields() filters out skipped fields
    let data_fields = data.fields();
    assert_eq!(data_fields.len(), 1, "Only non-skipped field remains");
    assert_eq!(
        data_fields[0].ident().as_ref().unwrap().to_string(),
        "a",
        "Expected remaining field to be 'a'"
    );
    assert!(
        data_fields[0].is_choice(),
        "Field 'a' should be marked as choice"
    );

    // Tuple variant: one skipped, one retained
    let tuple = variants
        .iter()
        .find(|v| v.ident().to_string() == "Tuple")
        .expect("Tuple variant present");
    assert!(matches!(tuple.style(), darling::ast::Style::Tuple));
    assert_eq!(tuple.all_fields().len(), 2, "All tuple fields parsed");
    assert_eq!(tuple.fields().len(), 1, "One tuple field was skipped");
    // For tuple fields, idents are None, so just ensure the remaining one is present
    assert!(tuple.fields()[0].ident().is_none());

    // Unit variant style check
    let unit = variants
        .iter()
        .find(|v| v.ident().to_string() == "Unit")
        .expect("Unit variant present");
    assert!(matches!(unit.style(), darling::ast::Style::Unit));
    assert!(unit.fields().is_empty());
}

#[test]
fn struct_display_strategy_default_is_fluent() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent] // no explicit display -> default for structs is FluentDisplay
        struct MyStruct {
            a: i32,
        }
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    assert_eq!(DisplayStrategy::from(&opts), DisplayStrategy::FluentDisplay);
    assert!(!opts.attr_args().is_this());
}

#[test]
fn struct_display_strategy_override_fluent_and_this() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(display = "fluent", this)]
        struct MyStruct {
            a: i32,
        }
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    assert_eq!(DisplayStrategy::from(&opts), DisplayStrategy::FluentDisplay);
    assert!(opts.attr_args().is_this());
}

#[test]
fn struct_kv_keys_parsing_and_field_skipping() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluentKv)]
        #[fluent_kv(keys = ["Error", "Notice"], this)]
        struct MyStruct {
            a: i32,
            #[fluent_kv(skip)]
            b: String,
        }
    };

    let opts = StructKvOpts::from_derive_input(&input).expect("StructKvOpts should parse");

    // ftl_enum_ident is <StructName>Ftl
    assert_eq!(opts.ftl_enum_ident().to_string(), "MyStructFtl");

    // keyyed_idents are <StructName><Key>Ftl
    let mut key_names: Vec<String> = opts
        .keyyed_idents()
        .into_iter()
        .map(|k| k.to_string())
        .collect();
    key_names.sort();
    assert_eq!(
        key_names,
        vec![
            "MyStructErrorFtl".to_string(),
            "MyStructNoticeFtl".to_string()
        ]
    );

    // fields() filters out skipped fields
    let fields = opts.fields();
    assert_eq!(fields.len(), 1);
    assert_eq!(fields[0].ident().as_ref().unwrap().to_string(), "a");

    // 'this' flag was set
    assert!(opts.attr_args().is_this());
}

#[test]
#[should_panic(expected = "Unexpected mode")]
fn enum_display_strategy_invalid_value_panics() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(display = "invalid")]
        enum MyEnum {
            Unit,
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    // This call should panic due to invalid display value
    let _ = DisplayStrategy::from(&opts);
}

#[test]
#[should_panic(expected = "Unexpected mode")]
fn struct_display_strategy_invalid_value_panics() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(display = "bad")]
        struct MyStruct {
            a: i32,
        }
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    // This call should panic due to invalid display value
    let _ = DisplayStrategy::from(&opts);
}

#[test]
fn struct_fluent_parsing() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(this)]
        struct MyStruct {
            a: i32,
            #[fluent(skip)]
            b: String,
            #[fluent(choice)]
            c: bool,
        }
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    insta::assert_debug_snapshot!(&opts);
}

#[test]
fn enum_choice_parsing() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluentChoice)]
        #[fluent_choice(serialize_all = "snake_case")]
        enum MyEnum {
            A,
            B,
        }
    };

    let opts = EnumChoiceOpts::from_derive_input(&input).expect("EnumChoiceOpts should parse");
    insta::assert_debug_snapshot!(&opts);
}
