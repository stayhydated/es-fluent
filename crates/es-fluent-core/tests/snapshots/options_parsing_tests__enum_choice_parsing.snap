---
source: crates/es-fluent-core/tests/options_parsing_tests.rs
expression: "&opts"
---
EnumChoiceOpts {
    ident: Ident {
        sym: MyEnum,
    },
    generics: Generics {
        lt_token: None,
        params: [],
        gt_token: None,
        where_clause: None,
    },
    data: Enum(
        [
            Ignored,
            Ignored,
        ],
    ),
    attr_args: EnumChoiceAttributeArgs {
        serialize_all: Some(
            "snake_case",
        ),
    },
}
